MobileCenter/MobileCenter/Internals/Application Delegate/MSAppDelegateForwarder.m

```diff
@@ -0,0 +1,277 @@
+
++ (NSHashTable<id<MSAppDelegate>> *)delegates {
+  return _delegates ?: (_delegates = [NSHashTable weakObjectsHashTable]);
+}
+
++ (void)setDelegates:(NSHashTable<id<MSAppDelegate>> *)delegates{
+  _delegates = delegates;
+}
+
++ (NSMutableSet<NSString *> *)selectorsToSwizzle {
+  return _selectorsToSwizzle ?: (_selectorsToSwizzle = [NSMutableSet new]);
+}
+
++ (NSMutableDictionary<NSString*, NSValue *> *)originalImplementations {
+  return _originalImplementations ?: (_originalImplementations = [NSMutableDictionary new]);
+}
+
++ (IMP)originalSetDelegateImp {
+  return _originalSetDelegateImp;
+}
+
++ (void)setOriginalSetDelegateImp:(IMP)originalSetDelegateImp {
+  _originalSetDelegateImp = originalSetDelegateImp;
+}
+
++ (BOOL)enabled {
+  @synchronized(self) {
+    return _enabled;
+  }
+}
+
++ (void)setEnabled:(BOOL)enabled {
+  @synchronized(self) {
+    _enabled = enabled;
+    if (!enabled) {
+      [self.delegates removeAllObjects];
+    }
+  }
+}
+
+#pragma mark - Delegates
+
++ (void)addDelegate:(id<MSAppDelegate>)delegate {
+  @synchronized(self) {
+    if (self.enabled) {
+      [self.delegates addObject:delegate];
+    }
+  }
+}
+
++ (void)removeDelegate:(id<MSAppDelegate>)delegate {
+  @synchronized(self) {
+    if (self.enabled) {
+      [self.delegates removeObject:delegate];
+    }
+  }
+}
+
+#pragma mark - Swizzling
+
++ (void)swizzleOriginalDelegate:(id<UIApplicationDelegate>) originalDelegate {
+  IMP originalImp = NULL;
+  Class delegateClass = [originalDelegate class];
+  SEL selector;
+  
+  // Swizzle all registered selectors.
+  for (NSString *selectorString in self.selectorsToSwizzle){
+    
+    // The same selector is used on both forwarder and delegate.
+    selector = NSSelectorFromString(selectorString);
+    originalImp = [self swizzleOriginalSelector:selector withCustomSelector:selector originalClass:delegateClass];
+    if (originalImp){
+      
+      // Save the original implementation for later use.
+      MSAppDelegateForwarder.originalImplementations[selectorString] = [NSValue valueWithBytes:&originalImp objCType:@encode(IMP)];
+    }
+  }
+  [self.selectorsToSwizzle removeAllObjects];
+}
+
++ (IMP)swizzleOriginalSelector:(SEL)originalSelector withCustomSelector:(SEL)customSelector originalClass:(Class)originalClass{
+  
+  // Replace original implementation
+  NSString * originalSelectorString = NSStringFromSelector(originalSelector);
+  Method originalMethod = class_getInstanceMethod(originalClass, originalSelector);
+  IMP customImp = class_getMethodImplementation(self, customSelector);
+  IMP originalImp = NULL;
+  BOOL methodAdded = NO;
+  
+  // Replace original implementation by the custom one.
+  if (originalMethod){
+    originalImp = method_setImplementation(originalMethod, customImp);
+  }
+  
+  /*
+   * The original class may not implement the selector (e.g.: optional method from protocol), 
+   * add the method to the original class and associate it to the custom implementation.
```

Comment:
In addition, I find this hard to read as the `else` comes after the comment. Maybe you can move the comment into the `else` block?